### PR TITLE
Add new `Style/RedundantRegexpWrapping` cop

### PR DESCRIPTION
**Pull request by koic**
This PR adds new `Style/RedundantRegexpConstructor` cop that checks for the instantiation of regexp using redundant `Regexp.new` or `Regexp.compile`.

```ruby
# bad
Regexp.new(/regexp/)
Regexp.compile(/regexp/)

# good
/regexp/
Regexp.new('regexp')
Regexp.compile('regexp')
```

Autocorrect replaces to regexp literal which is the simplest and fastest. There is speed benefits as well:

```ruby
% cat regexp_new.rb
#!/usr/local/bin/ruby

require 'benchmark/ips'

Benchmark.ips do |x|
  x.report('Regexp.new(/regexp/)') { Regexp.new(/regexp/) }
  x.report('Regexp.new("regexp")') { Regexp.new('regexp') }
  x.report('/regexp/')             { /regexp/ }

  x.compare!
end
```

```console
$ ruby regexp_new.rb
Warming up --------------------------------------
Regexp.new(/regexp/)    96.864k i/100ms
Regexp.new("regexp")    84.848k i/100ms
            /regexp/     2.755M i/100ms
Calculating -------------------------------------
Regexp.new(/regexp/)      1.011M (± 2.1%) i/s -      5.134M in   5.082604s
Regexp.new("regexp")    839.839k (± 3.2%) i/s -      4.242M in   5.057360s
            /regexp/     26.538M (± 9.6%) i/s -    132.263M in   5.055437s

Comparison:
            /regexp/: 26537811.8 i/s
Regexp.new(/regexp/):  1010544.6 i/s - 26.26x  slower
Regexp.new("regexp"):   839838.6 i/s - 31.60x  slower
```

And I think that it can be a general rule, so `Style` is better than `Performance`.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/style/redundant_regexp_wrapping.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # Checks for wrapping regexp literal with redundant `Regexp.new`.
```
Comment (bbatsov) [inline lib/rubocop/cop/style/redundant_regexp_wrapping.rb:6]: or `Regexp.compile`

Comment (koic) [reply lib/rubocop/cop/style/redundant_regexp_wrapping.rb]: Sure, I've updated it.

**Comment by bbatsov** [comment]
I'm wondering if `RedundantRegexpConstructor` might not be a better name for this, although perhaps there are other methods that might work similarly. 

I'm also wondering if we need a similar cop or should extend this cop to discourage the use of `Regexp.new` with string literals and no options, as it's better to just use regexp literals directly. I think we had some similar cops for array and hash literals.

**Comment by koic** [comment]
`RedundantRegexpConstructor` looks like a better name. I updated this PR with it.

> I'm also wondering if we need a similar cop or should extend this cop to discourage the use of `Regexp.new` with string literals and no options, as it's better to just use regexp literals directly.

Sure. This initial implementation accepts `Regexp.new` with string literals and options, but it could be replaced with a `/regexp/i` literal instead of `Regexp.new('regexp', Regexp::IGNORECASE)`. That decision may come later.

>  I think we had some similar cops for array and hash literals.

There was no cops to detect `Array.new([])` and `Hash.new({})`. So, other cops called `RedundantArrayConstructor` and `RedundantHashConstructor` could be implemented as well. I can implement them if you prefer :-)

**Comment by bbatsov** [comment]
> There was no cops to detect Array.new([]) and Hash.new({}). So, other cops called RedundantArrayConstructor and RedundantHashConstructor could be implemented as well. I can implement them if you prefer :-)

That would be great! 

> Sure. This initial implementation accepts Regexp.new with string literals and options, but it could be replaced with a /regexp/i literal instead of Regexp.new('regexp', Regexp::IGNORECASE). That decision may come later.

Yeah, we can polish the new cop down the road to cover such cases as well.

**Comment by koic** [comment]
`Hash.new` doesn't provide similar functionality, so it doesn't seem necessary to implement `RedundantHashConstructor`.
https://docs.ruby-lang.org/en/3.2/Hash.html#method-c-new

**Comment by FnControlOption** [comment]
Thoughts on [`String.new`](https://docs.ruby-lang.org/en/3.2/String.html#method-c-new)?

**Comment by koic** [comment]
It seems like that case would be handled by `Performance/UnfreezeString`.
https://docs.rubocop.org/rubocop-performance/cops_performance.html#performanceunfreezestring

https://github.com/rubocop/rubocop/pull/11886#issuecomment-1555940907

**Pull request merged**